### ISSUE TITLE
NCE or Negative Sampling

### ISSUE THREAD

**Issue by sachinruk** [keras-team/keras#4100]
Just wondering if there was a way to do NCE or negative sampling in keras. Tensorflow have these functions implemented in them. So if its not implemented in Keras, is there a way to wrap these functions?

**Comment by EderSantana** [comment]
are you talking about this: http://stackoverflow.com/questions/37671974/tensorflow-negative-sampling ?
 
if so, you can create your own custom loss functions and pass them to your compiler:
```python
def my_nce(y_true, y_pred):
  ...

model.compile('sgd', my_nce)
```

**Comment by sachinruk** [comment]
Yep. That's the one. I was hoping that keras had already implemented it. It's a necessity if you want to do a word2vec type model which has 100s+ of classes. So normal multinomial/ softmax won't do.

As for implementing it myself, I need to sample the `negative` classes, an integer from 0 to Nclasses. Can you do sampling **within** a keras loss function? (i.e. use np.random.choice?)

**Comment by EderSantana** [comment]
they discussed that on the issue. I think the tensorflow NCE code does that

**Comment by sachinruk** [comment]
I think you are missing the point of my question. I've already played around with tensorflow and used there version of NCE and negative sampling. Was wondering if there was a native way to do this on keras. Or at least if I can put this forward as a feature request on keras.

**Comment by Sandeep42** [comment]
Sorry for bumping up an old one. I am also interested in using NCE loss in Keras. Is there a way to go about it?

**Comment by RudrakshTuwani** [comment]
Same Question as above, I'm trying to implement Word2Vec in Keras and Softmax calculation is too expensive.

**Comment by ianupright** [comment]
I agree.  Chainer supports negative sampling and hierarchical softmax, built-in.  I'd like to see those features in Keras.

**Comment by andreidi** [comment]
I see this issue is closed however I haven't found any solutions yet... Any news on using sampled softmax in keras ?

**Comment by colpain** [comment]
Same here

**Comment by bicepjai** [comment]
nce available yet ?

**Comment by vsuthichai** [comment]
Bump.

Would love to know if anyone has rolled something custom to solve this.  I wouldn't mind re-using something that's working :)

**Comment by enockipp** [comment]
Same here， wait~

**Comment by jundl77** [comment]
+1

**Comment by cadama** [comment]
I did the [following](https://datascience.stackexchange.com/questions/28213/keras-negative-sampling-with-custom-layer) but apparently the gradient is not flowing.

**Comment by Apollo2Mars** [comment]
Hello everyone, does anyone know better NCE implementation code in keras? If anyone knows, could you stick a link?

**Comment by neopedro** [comment]
Anyone had sucsess with sampled softmax in keras?

**Comment by joelthchao** [comment]
@neopedro [WIP] My implementation for Keras (only tensorflow backend):
https://github.com/joelthchao/keras/tree/nce-loss-layer
```python
import numpy as np
from keras.layers import Dense, NCE, Input
from keras.models import Model

inputs = Input(shape=(4,))
target = Input(shape=(1,))
net = Dense(8)(inputs)
net = NCE(128, 32)([net, target])

model = Model(inputs=[inputs, target], outputs=net)
model.compile(optimizer='adam', loss=None)

x = np.random.rand(1000, 4)
y = np.random.randint(128, size=1000)
model.fit([x, y], None, batch_size=32, epochs=10)
model.predict([x, y])
model.evaluate([x, y], None)
```

**Comment by eric-haibin-lin** [comment]
Sampled softmax/NCE in gluon-nlp:
https://github.com/dmlc/gluon-nlp/blob/master/scripts/language_model/language_model.rst#large-scale-word-language-model
http://gluon-nlp.mxnet.io/api/model.html?highlight=isdense#gluonnlp.model.NCEDense

**Comment by nmishra** [comment]
> y = np.random.randint(128, size=1000)

How will this work in real world ? Usually, the target word should be a lookup from the embedding matrix to apply NCE. Hence, you need to find the +ve target word and -ve target word(s) around the center word to apply NCE loss.

**Comment by Santosh-Gupta** [comment]
Bump. Any solid solutions to getting sampled softmax in Keras?
